packages/react-router/src/Transitioner.tsx

```diff
@@ -87,74 +88,50 @@ export function Transitioner() {
   useLayoutEffect(() => {
     // The router was loading and now it's not
     if (previousIsLoading && !isLoading) {
-      const toLocation = router.state.location
-      const fromLocation = router.state.resolvedLocation
-      const pathChanged = fromLocation.pathname !== toLocation.pathname
-      const hrefChanged = fromLocation.href !== toLocation.href
-
       router.emit({
         type: 'onLoad', // When the new URL has committed, when the new matches have been loaded into state.matches
-        fromLocation,
-        toLocation,
-        pathChanged,
-        hrefChanged,
+        ...getLocationChangeInfo(router.state),
       })
     }
   }, [previousIsLoading, router, isLoading])
 
   useLayoutEffect(() => {
     // emit onBeforeRouteMount
     if (previousIsPagePending && !isPagePending) {
-      const toLocation = router.state.location
-      const fromLocation = router.state.resolvedLocation
-      const pathChanged = fromLocation.pathname !== toLocation.pathname
-      const hrefChanged = fromLocation.href !== toLocation.href
-
       router.emit({
         type: 'onBeforeRouteMount',
-        fromLocation,
-        toLocation,
-        pathChanged,
-        hrefChanged,
+        ...getLocationChangeInfo(router.state),
       })
     }
   }, [isPagePending, previousIsPagePending, router])
 
   useLayoutEffect(() => {
     // The router was pending and now it's not
     if (previousIsAnyPending && !isAnyPending) {
-      const toLocation = router.state.location
-      const fromLocation = router.state.resolvedLocation
-      const pathChanged = fromLocation.pathname !== toLocation.pathname
-      const hrefChanged = fromLocation.href !== toLocation.href
-
       router.emit({
         type: 'onResolved',
-        fromLocation,
-        toLocation,
-        pathChanged,
-        hrefChanged,
+        ...getLocationChangeInfo(router.state),
       })
 
       router.__store.setState((s) => ({
         ...s,
         status: 'idle',
         resolvedLocation: s.location,
       }))
-
-      if (typeof document !== 'undefined' && (document as any).querySelector) {
-        const hashScrollIntoViewOptions =
-          router.state.location.state.__hashScrollIntoViewOptions ?? true
-
-        if (hashScrollIntoViewOptions && router.state.location.hash !== '') {
-          const el = document.getElementById(router.state.location.hash)
-          if (el) {
-            el.scrollIntoView(hashScrollIntoViewOptions)
-          }
-        }
-      }
     }
   }, [isAnyPending, previousIsAnyPending, router])
 
   return null
 }
+
+export function getLocationChangeInfo(routerState: {
```

Comment:
Could this be exported out of the `router-core`?

Purely, so that we have less work when we eventually move more into the `router-core` package from `react-router` *(like the `Router` and `Route` classes, +others).

Reply:
Yep